Google.Api.Gax.Grpc/RpcExceptionExtensions.cs

```diff
@@ -0,0 +1,107 @@
+    /// <summary>
+    /// Utility extension methods to make it easier to retrieve extended error information from an <see cref="RpcException"/>.
+    /// </summary>
+    public static class RpcExceptionExtensions
+    {
+        // Visible for testing.
+        internal const string StatusDetailsTrailerName = "grpc-status-details-bin";
+
+        /// <summary>
+        /// Retrieves the <see cref="Status"/> message containing extended error information
+        /// from the trailers in an <see cref="RpcException"/>, if present.
+        /// </summary>
+        /// <param name="ex">The RPC exception to retrieve details from. Must not be null.</param>
+        /// <returns>The <see cref="Status"/> message specified in the exception, or null
+        /// if there is no such information.</returns>
+        public static Rpc.Status GetRpcStatus(this RpcException ex) =>
+            DecodeTrailer(GetTrailer(ex, StatusDetailsTrailerName), Status.Parser);
+
+        /// <summary>
+        /// Retrieves the <see cref="BadRequest"/> message containing extended error information
+        /// from the trailers in an <see cref="RpcException"/>, if present.
+        /// </summary>
+        /// <param name="ex">The RPC exception to retrieve details from. Must not be null.</param>
+        /// <returns>The <see cref="BadRequest"/> message specified in the exception, or null
+        /// if there is no such information.</returns>
+        public static BadRequest GetBadRequest(this RpcException ex) => GetStatusDetail<BadRequest>(ex);
+
+        /// <summary>
+        /// Retrieves the <see cref="ErrorInfo"/> message containing extended error information
+        /// from the trailers in an <see cref="RpcException"/>, if present.
+        /// </summary>
+        /// <param name="ex">The RPC exception to retrieve details from. Must not be null.</param>
+        /// <returns>The <see cref="ErrorInfo"/> message specified in the exception, or null
+        /// if there is no such information.</returns>
+        public static ErrorInfo GetErrorInfo(this RpcException ex) => GetStatusDetail<ErrorInfo>(ex);
+
+        /// <summary>
+        /// Retrieves the error details of type <typeparamref name="T"/> from the <see cref="Status"/>
+        /// message associated with an <see cref="RpcException"/>, if any.
+        /// </summary>
+        /// <typeparam name="T">The message type to decode from within the error details.</typeparam>
+        /// <param name="ex">The RPC exception to retrieve details from. Must not be null.</param>
+        /// <returns></returns>
+        public static T GetStatusDetail<T>(this RpcException ex) where T : class, IMessage<T>, new()
+        {
+            var status = GetRpcStatus(ex);
+            if (status is null)
+            {
+                return null;
+            }
+            var expectedName = new T().Descriptor.FullName;
+            var any = status.Details.FirstOrDefault(a => Any.GetTypeName(a.TypeUrl) == expectedName);
```

Comment:
If we can have many of these (not sure that's the case though), maybe add a version of this method that returns all of them? `IEnumerable<T> GetStatusDetails`?

Reply:
While in theory it would be possible to have multiple values for the same message type, I think it vanishingly unlikely. I suggest we wait until it's requested - expecting that it never will be :)

(Users can always do it themselves, of course.)